# Data Understanding
---

**Name:** Sapto Mukti Handoyo

**Dataset Source:**
https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce
\
**Dataset License (from Kaggle Website):** https://creativecommons.org/licenses/by-nc-sa/4.0 (CC BY-NC-SA 4.0)

# Import Package

In [1]:
import pandas as pd
import numpy as np

# Dataset Preparation & Selection of Primary Key (PK)

## 1. Olist Orders Dataset

In [2]:
# Read CSV File
orders = pd.read_csv("C:/Users/Sapto/Documents/projects/data/raw/olist_orders_dataset.csv")

In [3]:
# Dataset Size
orders.shape

(99441, 8)

Total Rows: 99441\
Total Column: 8

In [4]:
# Column Name
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')

In [5]:
# Dataset Snippet
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

ID Column:
- `order_id`
- `customer_id`

Date Column:
- `order_purchase_timestamp`
- `order_approved_at`
- `order_delivered_carrier_date`
- `order_delivered_customer_date`
- `order_estimated_delivery_date`

In [6]:
# Column Types
orders.dtypes

order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

In [7]:
# Check NULL (Snippet)
orders.isna()

order_id  customer_id  order_status  order_purchase_timestamp  \
0         False        False         False                     False   
1         False        False         False                     False   
2         False        False         False                     False   
3         False        False         False                     False   
4         False        False         False                     False   
...         ...          ...           ...                       ...   
99436     False        False         False                     False   
99437     False        False         False                     False   
99438     False        False         False                     False   
99439     False        False         False                     False   
99440     False        False         False                     False   

       order_approved_at  order_delivered_carrier_date  \
0                  False                         False   
1                  False                         False   
2                  False                         False   
3                  False                         False   
4                  False                         False   
...                  ...                           ...   
99436              False                         False   
99437              False                         False   
99438              False                         False   
99439              False                         False   
99440              False                         False   

       order_delivered_customer_date  order_estimated_delivery_date  
0                              False                          False  
1                              False                          False  
2                              False                          False  
3                              False                          False  
4                              False                          False  
...                              ...                            ...  
99436                          False                          False  
99437                          False                          False  
99438                          False                          False  
99439                          False                          False  
99440                          False                          False  

[99441 rows x 8 columns]

In [8]:
# Check NULL Count
orders.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

Columns that contain Null Value:
- order_approved_at:                 160 Null Values
- order_delivered_carrier_date:     1783 Null Values
- order_delivered_customer_date:    2965 Null Values

In [9]:
orders.duplicated('order_id').sum() == 0
# or we can use orders['order_id'].nunique() == len(orders)
# or we can use orders['order_id'].value_counts()
# or we can use orders['order_id'].is_unique

np.True_

In [10]:
# Check Null Value in PK
orders['order_id'].isna().sum()

np.int64(0)

- 1 row represents 1 order (`order_id`).
- `order_id` is unique & not null types of column.
- So, `order_id` can be chosen as a PK.

## 2. Olist Order Items Dataset

In [11]:
# Read CSV File
order_items = pd.read_csv("C:/Users/Sapto/Documents/projects/data/raw/olist_order_items_dataset.csv")

In [12]:
# Dataset Size
order_items.shape

(112650, 7)

Total Rows: 112650\
Total Columns: 7

In [13]:
# Column Name
order_items.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='object')

In [14]:
# Dataset Snippet
order_items.head(10)

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   
5  00048cc3ae777c65dbb7d2a0634bc1ea              1   
6  00054e8431b9d7675808bcb819fb4a32              1   
7  000576fe39319847cbb9d288c5617fa6              1   
8  0005a1a1728c9d785b8e2b08b904576c              1   
9  0005f50442cb953dcd1d21e1fb923495              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
5  ef92defde845ab8450f9d70c526ef70f  6426d21aca402a131fc0a5d0960a3c90   
6  8d4f2bb7e93e6710a28f34fa83ee7d28  7040e82f899a04d1b434b795a43b4617   
7  557d850972a7d6f792fd18ae1400d9b6  5996cddab893a4652a15592fb58ab8db   
8  310ae3c140ff94b03219ad0adc3c778f  a416b6a846a11724393025641d4edd5e   
9  4535b0e1091c278dfd193e5a1d63b39f  ba143b05f0110f0dc71ad71b4466ce92   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14  
5  2017-05-23 03:55:27   21.90          12.69  
6  2017-12-14 12:10:31   19.90          11.85  
7  2018-07-10 12:30:45  810.00          70.75  
8  2018-03-26 18:31:29  145.95          11.65  
9  2018-07-06 14:10:56   53.99          11.40

ID Column:
- `order_id`
- `order_item_id`
- `product_id`
- `seller_id`

Date Column:
- `shipping_limit_date`

In [15]:
# Column Types
order_items.dtypes

order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

In [16]:
# Check NULL (Snippet)
order_items.isna()

order_id  order_item_id  product_id  seller_id  shipping_limit_date  \
0          False          False       False      False                False   
1          False          False       False      False                False   
2          False          False       False      False                False   
3          False          False       False      False                False   
4          False          False       False      False                False   
...          ...            ...         ...        ...                  ...   
112645     False          False       False      False                False   
112646     False          False       False      False                False   
112647     False          False       False      False                False   
112648     False          False       False      False                False   
112649     False          False       False      False                False   

        price  freight_value  
0       False          False  
1       False          False  
2       False          False  
3       False          False  
4       False          False  
...       ...            ...  
112645  False          False  
112646  False          False  
112647  False          False  
112648  False          False  
112649  False          False  

[112650 rows x 7 columns]

In [17]:
# Check NULL Count
order_items.isna().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [18]:
order_items.duplicated('order_id').sum() == 0
# or we can use order_items['order_id'].nunique() == len(order_items)
# or we can use order_items['order_id'].value_counts()
# or we can use order_items['order_id'].is_unique

np.False_

In [19]:
order_items.duplicated('order_item_id').sum() == 0
# or we can use order_items['order_item_id'].nunique() == len(order_items)
# or we can use order_items['order_item_id'].value_counts()
# or we can use order_items['order_item_id'].is_unique

np.False_

In [20]:
order_items['order_id'].isna().sum()

np.int64(0)

In [21]:
order_items['order_item_id'].isna().sum()

np.int64(0)

In [22]:
order_items[order_items['order_item_id'].isin([1, 2, 3, 4])].groupby('order_item_id', as_index=False).count()[['order_id', 'order_item_id']]

order_id  order_item_id
0     98666              1
1      9803              2
2      2287              3
3       965              4

In [23]:
order_items[order_items['order_id'].isin(['02a065131a2d2b72b45e2c63135606ad', '03aba68b07658f28f29612641f08d4ba'])].sort_values(by='order_id', ascending=True)[['order_id', 'order_item_id']]

order_id  order_item_id
1132  02a065131a2d2b72b45e2c63135606ad              1
1133  02a065131a2d2b72b45e2c63135606ad              2
1134  02a065131a2d2b72b45e2c63135606ad              3
1135  02a065131a2d2b72b45e2c63135606ad              4
1136  02a065131a2d2b72b45e2c63135606ad              5
1593  03aba68b07658f28f29612641f08d4ba              1
1594  03aba68b07658f28f29612641f08d4ba              2
1595  03aba68b07658f28f29612641f08d4ba              3
1596  03aba68b07658f28f29612641f08d4ba              4
1597  03aba68b07658f28f29612641f08d4ba              5
1598  03aba68b07658f28f29612641f08d4ba              6

Further inspection:
- `order_item_id` is the item sequence number IN ONE ORDER. This number is reset for each order.
- `order_item_id` is only unique within one order, not globally unique in the table.
- `order_id` is also not a unique column.
- So, we can't use `order_item_id` and `order_id` as a primary key.
- We need another way to determine a primary key for this table.

Ideas:
- A primary key is determined by grain.
- 'Grain' means what is represented by 1 row in the table.
- We see that 1 row represents 1 item within an order (`order_item_id` is only unique within ONE ORDER).
- In this case, grain = 1 item within an order.
- The pairs of `order_id` and `order_item_id` may form a primary key. We observe that these pairs are unique values:
  - `('order_id1', order_item_id1)`
  - `('order_id1', order_item_id2)`
  - `('order_id1', order_item_id3)`
  - `('order_id2', order_item_id1)`
  - `('order_id2', order_item_id2)`
  - and so on.

In [24]:
# Check Uniqueness in PK
order_items.duplicated(subset=['order_id', 'order_item_id']).sum() == 0

np.True_

In [25]:
# Check Null Value in PK
order_items[['order_id', 'order_item_id']].isna().sum()

order_id         0
order_item_id    0
dtype: int64

Thus, we can use `(order_id, order_item_id)` as a **composite** primary key.

## 3. Olist Order Payments Dataset

In [26]:
# Read CSV File
order_payments = pd.read_csv("C:/Users/Sapto/Documents/projects/data/raw/olist_order_payments_dataset.csv")

In [27]:
# Dataset Size
order_payments.shape

(103886, 5)

Total Rows: 103886\
Total Columns: 5

In [28]:
# Column Name
order_payments.columns

Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')

In [29]:
# Dataset Snippet
order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

ID Column: `order_id`

In [30]:
# Column Types
order_payments.dtypes

order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

In [31]:
# Check NULL (Snippet)
order_payments.isna()

order_id  payment_sequential  payment_type  payment_installments  \
0          False               False         False                 False   
1          False               False         False                 False   
2          False               False         False                 False   
3          False               False         False                 False   
4          False               False         False                 False   
...          ...                 ...           ...                   ...   
103881     False               False         False                 False   
103882     False               False         False                 False   
103883     False               False         False                 False   
103884     False               False         False                 False   
103885     False               False         False                 False   

        payment_value  
0               False  
1               False  
2               False  
3               False  
4               False  
...               ...  
103881          False  
103882          False  
103883          False  
103884          False  
103885          False  

[103886 rows x 5 columns]

In [32]:
# Check NULL Count
order_payments.isna().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [33]:
order_payments.duplicated('order_id').sum() == 0
# or we can use order_payments.nunique['payment_sequential'] == len(order_payments)
# or we can use order_payments['order_id'].value_counts()
# or we can use order_payments['order_id'].is_unique

np.False_

In [34]:
order_payments.duplicated('payment_sequential').sum() == 0
# or we can use order_payments['payment_sequential'].nunique() == len(order_payments)
# or we can use order_payments['payment_sequential'].value_counts()
# or we can use order_payments['payment_sequential'].is_unique

np.False_

In [35]:
order_payments['order_id'].isna().sum()

np.int64(0)

In [36]:
order_payments['payment_sequential'].isna().sum()

np.int64(0)

In [37]:
order_payments[order_payments['payment_sequential'].isin([1,2,3,4])].groupby('payment_sequential', as_index=False).count()[['order_id', 'payment_sequential']]

order_id  payment_sequential
0     99360                   1
1      3039                   2
2       581                   3
3       278                   4

In [38]:
order_payments[order_payments['order_id'].isin(['00c405bd71187154a7846862f585a9d4','fefbe15ebcd87ab3fb8577e635a8b31c'])].sort_values(by=['order_id', 'payment_sequential'], ascending=[True, True])[['order_id', 'payment_sequential']]

order_id  payment_sequential
67123   00c405bd71187154a7846862f585a9d4                   1
54653   00c405bd71187154a7846862f585a9d4                   2
60979   00c405bd71187154a7846862f585a9d4                   3
76930   00c405bd71187154a7846862f585a9d4                   4
52049   00c405bd71187154a7846862f585a9d4                   5
26391   00c405bd71187154a7846862f585a9d4                   6
20594   00c405bd71187154a7846862f585a9d4                   7
101581  fefbe15ebcd87ab3fb8577e635a8b31c                   1
25698   fefbe15ebcd87ab3fb8577e635a8b31c                   2
6904    fefbe15ebcd87ab3fb8577e635a8b31c                   3
68121   fefbe15ebcd87ab3fb8577e635a8b31c                   4
87435   fefbe15ebcd87ab3fb8577e635a8b31c                   5
73248   fefbe15ebcd87ab3fb8577e635a8b31c                   6

Further inspection:
- `payment_sequential` is the item sequence number IN ONE ORDER. This number is reset for each order.
- `payment_sequential` is only unique within one order, not globally unique in the table.
- `order_id` is also not a unique column.
- So, we can't use `payment_sequential` and `order_id` as a primary key.
- We need another way to determine a primary key for this table.

Ideas:
- A primary key is determined by grain.
- 'Grain' means what is represented by 1 row in the table.
- We see that 1 row represents 1 payment within an order (`payment_sequential` is only unique within ONE ORDER).
- In this case, grain = 1 payment within an order.
- The pairs of `order_id` and `payment_sequential` may form a primary key. We observe that these pairs are unique values:
  - `('order_id1', payment_sequential1)`
  - `('order_id1', payment_sequential2)`
  - `('order_id1', payment_sequential3)`
  - `('order_id2', payment_sequential1)`
  - `('order_id2', payment_sequential2)`
  - and so on.

In [39]:
# Check Uniqueness in PK
order_payments.duplicated(subset=['order_id', 'payment_sequential']).sum() == 0

np.True_

In [40]:
# Check Null Value in PK
order_payments[['order_id', 'payment_sequential']].isna().sum()

order_id              0
payment_sequential    0
dtype: int64

Thus, we can use `(order_id, payment_sequential)` as a **composite** primary key.

## 4. Olist Order Reviews Dataset

In [41]:
# Read CSV File
order_reviews = pd.read_csv("C:/Users/Sapto/Documents/projects/data/raw/olist_order_reviews_dataset.csv")

In [42]:
# Dataset Size
order_reviews.shape

(99224, 7)

Total Rows: 99224\
Total Columns: 7

In [43]:
# Column Name
order_reviews.columns

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='object')

In [44]:
# Dataset Snippet
order_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

ID Column:
- `review_id`
- `order_id`

Date Column:
- `review_creation_date`
- `review_answer_timestamp`

In [45]:
# Column Types
order_reviews.dtypes

review_id                  object
order_id                   object
review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
dtype: object

In [46]:
# Check NULL (Snippet)
order_reviews.isna()

review_id  order_id  review_score  review_comment_title  \
0          False     False         False                  True   
1          False     False         False                  True   
2          False     False         False                  True   
3          False     False         False                  True   
4          False     False         False                  True   
...          ...       ...           ...                   ...   
99219      False     False         False                  True   
99220      False     False         False                  True   
99221      False     False         False                  True   
99222      False     False         False                  True   
99223      False     False         False                  True   

       review_comment_message  review_creation_date  review_answer_timestamp  
0                        True                 False                    False  
1                        True                 False                    False  
2                        True                 False                    False  
3                       False                 False                    False  
4                       False                 False                    False  
...                       ...                   ...                      ...  
99219                    True                 False                    False  
99220                    True                 False                    False  
99221                   False                 False                    False  
99222                    True                 False                    False  
99223                   False                 False                    False  

[99224 rows x 7 columns]

In [47]:
# Check NULL Count
order_reviews.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

Columns that contain Null Value:
- review_comment_title:       87656 Null Values
- review_comment_message:     58247 Null Values

In [48]:
order_reviews['review_id'].duplicated().sum() == 0
# or we can use order_reviews['order_id'].nunique() == len(order_reviews)
# or we can use order_reviews['order_id'].value_counts()
# or we can use order_reviews['order_id'].is_unique

np.False_

In [49]:
order_reviews['order_id'].duplicated().sum() == 0
# or we can use order_reviews['order_id'].nunique() == len(order_reviews)
# or we can use order_reviews['order_id'].value_counts()
# or we can use order_reviews['order_id'].is_unique

np.False_

In [50]:
order_reviews['review_id'].isna().sum()

np.int64(0)

In [51]:
order_reviews['order_id'].isna().sum()

np.int64(0)

In [52]:
order_reviews[order_reviews.duplicated(subset=['review_id'], keep=False)].sort_values(by='review_id').head(4)

review_id                          order_id  \
46678  00130cbe1f9d422698c812ed8ded1919  dfcdfc43867d1c1381bfaf62d6b9c195   
29841  00130cbe1f9d422698c812ed8ded1919  04a28263e085d399c97ae49e0b477efa   
90677  0115633a9c298b6a98bcbe4eee75345f  78a4201f58af3463bdab842eea4bc801   
63193  0115633a9c298b6a98bcbe4eee75345f  0c9850b2c179c1ef60d2855e2751d1fa   

       review_score review_comment_title  \
46678             1                  NaN   
29841             1                  NaN   
90677             5                  NaN   
63193             5                  NaN   

                                  review_comment_message review_creation_date  \
46678  O cartucho "original HP" 60XL não é reconhecid...  2018-03-07 00:00:00   
29841  O cartucho "original HP" 60XL não é reconhecid...  2018-03-07 00:00:00   
90677                                                NaN  2017-09-21 00:00:00   
63193                                                NaN  2017-09-21 00:00:00   

      review_answer_timestamp  
46678     2018-03-20 18:08:23  
29841     2018-03-20 18:08:23  
90677     2017-09-26 03:27:47  
63193     2017-09-26 03:27:47

In [53]:
order_reviews[order_reviews.duplicated(subset=['order_id'], keep=False)].sort_values(by='order_id').head(4)

review_id                          order_id  \
25612  89a02c45c340aeeb1354a24e7d4b2c1e  0035246a40f520710769010f752e7507   
22423  2a74b0559eb58fc1ff842ecc999594cb  0035246a40f520710769010f752e7507   
22779  ab30810c29da5da8045216f0f62652a2  013056cfe49763c6f66bda03396c5ee3   
68633  73413b847f63e02bc752b364f6d05ee9  013056cfe49763c6f66bda03396c5ee3   

       review_score review_comment_title  \
25612             5                  NaN   
22423             5                  NaN   
22779             5                  NaN   
68633             4                  NaN   

                                  review_comment_message review_creation_date  \
25612                                                NaN  2017-08-29 00:00:00   
22423  Estou acostumada a comprar produtos pelo barat...  2017-08-25 00:00:00   
22779                                                NaN  2018-02-22 00:00:00   
68633                                                NaN  2018-03-04 00:00:00   

      review_answer_timestamp  
25612     2017-08-30 01:59:12  
22423     2017-08-29 21:45:57  
22779     2018-02-23 12:12:30  
68633     2018-03-05 17:02:00

Further inspection:
- `order_id` and `review_id` are not unique columns.
- 1 `review_id` can appear in 2 different `order_id`, and vice versa.
- So, we can't use `review_id` and `order_id` as a primary key.
- We need another way to determine a primary key for this table.

Ideas:
- A primary key is determined by grain.
- 'Grain' means what is represented by 1 row in the table.
- We see that 1 row represents a relationship between a review and an order.
- The pairs of `order_id` and `review_id` may form a primary key. We observe that these pairs are unique values:
  - `('order_id1', review_id1)`
  - `('order_id1', review_id2)`
  - `('order_id2', review_id3)`
  - `('order_id2', review_id4)`
  - and so on.

In [54]:
# Check Uniqueness in PK
order_reviews.duplicated(subset=['order_id', 'review_id']).sum() == 0

np.True_

In [55]:
# Check Null Value in PK
order_reviews[['order_id', 'review_id']].isna().sum()

order_id     0
review_id    0
dtype: int64

Thus, we can use `(order_id, review_id)` as a **composite** primary key.

## 5. Olist Customers Dataset

In [56]:
# Read CSV File
customers = pd.read_csv("C:/Users/Sapto/Documents/projects/data/raw/olist_customers_dataset.csv")

In [57]:
# Dataset Size
customers.shape

(99441, 5)

Total Rows: 99441\
Total Columns: 5

In [58]:
# Column Name
customers.columns

Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='object')

In [59]:
# Dataset Snippet
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

ID Column:
- `customer_id`
- `customer_unique_id`

In [60]:
# Column Types
customers.dtypes

customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

In [61]:
# Check NULL (Snippet)
customers.isna()

customer_id  customer_unique_id  customer_zip_code_prefix  \
0            False               False                     False   
1            False               False                     False   
2            False               False                     False   
3            False               False                     False   
4            False               False                     False   
...            ...                 ...                       ...   
99436        False               False                     False   
99437        False               False                     False   
99438        False               False                     False   
99439        False               False                     False   
99440        False               False                     False   

       customer_city  customer_state  
0              False           False  
1              False           False  
2              False           False  
3              False           False  
4              False           False  
...              ...             ...  
99436          False           False  
99437          False           False  
99438          False           False  
99439          False           False  
99440          False           False  

[99441 rows x 5 columns]

In [62]:
# Check NULL Count
customers.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [63]:
customers['customer_id'].duplicated().sum() == 0
# customers['customer_id'].nunique() == len(customers)
# customers.value_counts(['customer_id'])
# customers['customer_id'].is_unique

np.True_

In [64]:
customers['customer_unique_id'].duplicated().sum() == 0
# customers['customer_unique_id'].nunique() == len(customers)
# customers.value_counts(['customer_unique_id'])
# customers['customer_unique_id'].is_unique

np.False_

In [65]:
customers[['customer_id', 'customer_unique_id']].isna().sum()

customer_id           0
customer_unique_id    0
dtype: int64

- `customer_id` and `customer_unique_id` are not null columns.
- `customer_id` is a unique column, but `customer_unique_id` isn't.
- 1 row may represent a customer snapshot (customer_id level).
- In this case, `customer_unique_id` may represents real customer (real person).
- 1 real customer (`customer_unique_id`) may have 2 or more different `customer_id`.
- Maybe, 1 real customer (1 person) can: change address/personal data, have one or more transactions/orders, etc., so it is reasonable if 1 real customer can have 2 or more `customer_id`.
- Thus, we can use `customer_id` as a PK.

## 6. Olist Products Dataset

In [66]:
# Read CSV File
products = pd.read_csv("C:/Users/Sapto/Documents/projects/data/raw/olist_products_dataset.csv")

In [67]:
# Dataset Size
products.shape

(32951, 9)

Total Rows: 32951\
Total Columns: 9

In [68]:
# Column Name
products.columns

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='object')

In [69]:
# Dataset Snippet
products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

ID Column: `product_id`

In [70]:
# Column Types
products.dtypes

product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

In [71]:
# Check NULL (Snippet)
products.isnull()

product_id  product_category_name  product_name_lenght  \
0           False                  False                False   
1           False                  False                False   
2           False                  False                False   
3           False                  False                False   
4           False                  False                False   
...           ...                    ...                  ...   
32946       False                  False                False   
32947       False                  False                False   
32948       False                  False                False   
32949       False                  False                False   
32950       False                  False                False   

       product_description_lenght  product_photos_qty  product_weight_g  \
0                           False               False             False   
1                           False               False             False   
2                           False               False             False   
3                           False               False             False   
4                           False               False             False   
...                           ...                 ...               ...   
32946                       False               False             False   
32947                       False               False             False   
32948                       False               False             False   
32949                       False               False             False   
32950                       False               False             False   

       product_length_cm  product_height_cm  product_width_cm  
0                  False              False             False  
1                  False              False             False  
2                  False              False             False  
3                  False              False             False  
4                  False              False             False  
...                  ...                ...               ...  
32946              False              False             False  
32947              False              False             False  
32948              False              False             False  
32949              False              False             False  
32950              False              False             False  

[32951 rows x 9 columns]

In [72]:
# Check NULL Count
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

Columns that contain Null Value:
- product_category_name         610 Null Values
- product_name_lenght           610 Null Values
- product_description_lenght    610 Null Values
- product_photos_qty            610 Null Values
- product_weight_g                2 Null Values
- product_length_cm               2 Null Values
- product_height_cm               2 Null Values
- product_width_cm                2 Null Values

In [73]:
products['product_id'].duplicated().sum() == 0
# products['product_id'].nunique() == len(products)
# products.value_counts(['product_id'])
# products['product_id'].is_unique

np.True_

In [74]:
products['product_id'].isna().sum()

np.int64(0)

- `product_id` is a unique and not null columns.
- 1 row represents only 1 product.
- Thus, we can use `product_id` as a PK.

## 7. Olist Sellers Dataset

In [75]:
# Read CSV File
sellers = pd.read_csv("C:/Users/Sapto/Documents/projects/data/raw/olist_sellers_dataset.csv")

In [76]:
# Dataset Size
sellers.shape

(3095, 4)

Total Rows: 3095\
Total Columns: 4

In [77]:
# Column Name
sellers.columns

Index(['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state'], dtype='object')

In [78]:
# Dataset Snippet
sellers.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

ID Column: `seller_id`

In [79]:
# Column Types
sellers.dtypes

seller_id                 object
seller_zip_code_prefix     int64
seller_city               object
seller_state              object
dtype: object

In [80]:
# Check NULL (Snippet)
sellers.isnull()

seller_id  seller_zip_code_prefix  seller_city  seller_state
0         False                   False        False         False
1         False                   False        False         False
2         False                   False        False         False
3         False                   False        False         False
4         False                   False        False         False
...         ...                     ...          ...           ...
3090      False                   False        False         False
3091      False                   False        False         False
3092      False                   False        False         False
3093      False                   False        False         False
3094      False                   False        False         False

[3095 rows x 4 columns]

In [81]:
# Check NULL Count
sellers.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [82]:
sellers['seller_id'].duplicated().sum() == 0
# sellers['seller_id'].nunique() == len(sellers)
# sellers.value_counts(['seller_id'])

np.True_

In [83]:
sellers['seller_id'].isna().sum()

np.int64(0)

- `seller_id` is a unique and not null columns.
- 1 row represents only 1 seller.
- Thus, we can use `seller_id` as a PK.

## 8. Olist Geolocation Dataset

In [84]:
# Read CSV File
geolocation = pd.read_csv("C:/Users/Sapto/Documents/projects/data/raw/olist_geolocation_dataset.csv")

In [85]:
# Dataset Size
geolocation.shape

(1000163, 5)

Total Rows: 1000163\
Total Columns: 5

In [86]:
# Column Name
geolocation.columns

Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='object')

In [87]:
# Dataset Snippet
geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [88]:
# Column Types
geolocation.dtypes

geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                object
geolocation_state               object
dtype: object

In [89]:
# Check NULL (Snippet)
geolocation.isnull()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                              False            False            False   
1                              False            False            False   
2                              False            False            False   
3                              False            False            False   
4                              False            False            False   
...                              ...              ...              ...   
1000158                        False            False            False   
1000159                        False            False            False   
1000160                        False            False            False   
1000161                        False            False            False   
1000162                        False            False            False   

         geolocation_city  geolocation_state  
0                   False              False  
1                   False              False  
2                   False              False  
3                   False              False  
4                   False              False  
...                   ...                ...  
1000158             False              False  
1000159             False              False  
1000160             False              False  
1000161             False              False  
1000162             False              False  

[1000163 rows x 5 columns]

In [90]:
# Check NULL Count
geolocation.isnull().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [91]:
geolocation.duplicated(
    subset=[
        'geolocation_zip_code_prefix',
        'geolocation_lat',
        'geolocation_lng',
        'geolocation_city',
        'geolocation_state'
    ],
    keep=False
).sum()

np.int64(390005)

- Duplicates occur in all columns in the corresponding rows for a total of 390005 rows.
- Therefore, no primary key can be created in this table.
- This table can be used as a reference or lookup table for location mapping.

## 9. Product Category Name Translation Dataset

In [93]:
# Read CSV File
category_name = pd.read_csv("C:/Users/Sapto/Documents/projects/data/raw/product_category_name_translation.csv")

In [94]:
# Dataset Size
category_name.shape

(71, 2)

Total Rows: 71\
Total Columns: 2

In [95]:
# Column Name
category_name.columns

Index(['product_category_name', 'product_category_name_english'], dtype='object')

In [96]:
# Dataset Snippet
category_name.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [97]:
# Column Types
category_name.dtypes

product_category_name            object
product_category_name_english    object
dtype: object

In [98]:
# Check NULL (Snippet)
category_name.isnull()

product_category_name  product_category_name_english
0                   False                          False
1                   False                          False
2                   False                          False
3                   False                          False
4                   False                          False
..                    ...                            ...
66                  False                          False
67                  False                          False
68                  False                          False
69                  False                          False
70                  False                          False

[71 rows x 2 columns]

In [99]:
# Check NULL Count
category_name.isnull().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

In [100]:
category_name['product_category_name'].duplicated().sum() == 0
# sellers['seller_id'].nunique() == len(sellers)
# sellers.value_counts(['seller_id'])

np.True_

In [101]:
category_name['product_category_name'].isna().sum()

np.int64(0)

- `product_category_name` is a unique and not null columns.
- 1 row represents only 1 product translation.
- Thus, we can use `product_category_name` as a PK.

**Dataset Variable Names & Their PK Summaries:**
- orders
  - PK: order_id
  - Other ID: customer_id
- order_items
  - PK: (order_id, order_item_id)
  - Other ID: product_id, seller_id
- order_payments
  - PK: (order_id, payment_sequential)
- order_reviews
  - PK: (order_id, review_id)
- customers
  - PK: customer_id
  - Other ID: customer_unique_id, customer_zip_code_prefix
- products
  - PK: product_id
  - Other ID: product_category_name
- sellers
  - PK: seller_id
  - Other ID: seller_zip_code_prefix
- geolocation
  - No PK
  - Important Column: geolocation_zip_code_prefix
- category_name
  - PK: product_category_name

# Selection of Foreign Key (FK)

## 1. Olist Orders Dataset

In [102]:
orders['customer_id'].isin(customers['customer_id']).mean()

np.float64(1.0)

FK: `customer_id`

## 2. Olist Order Items Dataset

In [103]:
order_items['product_id'].isin(products['product_id']).mean()

np.float64(1.0)

In [104]:
order_items['seller_id'].isin(sellers['seller_id']).mean()

np.float64(1.0)

In [105]:
order_items['order_id'].isin(orders['order_id']).mean()

np.float64(1.0)

FK: `product_id`, `seller_id`, `order_id`

## 3. Olist Order Payments Dataset

In [106]:
order_payments['order_id'].isin(orders['order_id']).mean()

np.float64(1.0)

FK: `order_id`

## 4. Olist Order Reviews Dataset

In [107]:
order_reviews['order_id'].isin(orders['order_id']).mean()

np.float64(1.0)

FK: `order_id`

## 5. Olist Customers Dataset

In [108]:
customers['customer_zip_code_prefix'].isin(geolocation['geolocation_zip_code_prefix']).mean()

np.float64(0.9972043724419505)

mean < 1: There is record in the `customer_zip_code_prefix` column in the `customers` dataset (child table) that is not in the `geolocation_zip_code_prefix` column in the `geolocation` dataset (parent table) (we called this record as an **orphan record**). Here is a list of that record:

In [109]:
customers.loc[
    ~customers['customer_zip_code_prefix'].isin(geolocation['geolocation_zip_code_prefix']),
    'customer_zip_code_prefix'
].unique()

array([72300, 11547, 64605, 72465,  7729, 72904, 35408, 78554, 73369,
        8980, 29949, 65137, 28655, 73255, 28388,  6930, 71676, 64047,
       61906, 83210, 71919, 36956, 35242, 72005, 29718, 41347, 70324,
       70686, 72341, 12332, 70716, 71905, 75784, 73082, 71884, 71574,
       72238, 71996, 76968, 71975, 72595, 72017, 72596, 67105, 25840,
       72002, 72821, 85118, 25919, 95853, 72583, 68511, 70701, 71591,
       72535, 95572, 73090, 72242, 86135, 70316, 73091, 41098, 58734,
       73310, 71810, 72280,  7430, 73081, 70333, 72268, 35104, 72455,
       72237, 17390, 76897, 84623, 70702, 72760, 73088, 29196, 36596,
       57254, 71995, 73093, 75257, 48504, 83843, 62625, 37005, 73401,
       49870, 13307, 28617, 73402, 56327, 71976, 72587, 85958, 19740,
       77404, 44135, 28120, 72863, 87323, 87511, 72440, 72243, 65830,
       71261, 28575,  2140, 71551, 72023, 28160, 55027, 43870, 94370,
       38710, 42716, 36248, 71593, 71953, 72549, 72457, 56485, 71590,
       93602,  7412,

FK: `customer_zip_code_prefix`

## 6. Olist Products Dataset

In [110]:
products['product_category_name'].isin(category_name['product_category_name']).mean()

np.float64(0.9810931382962581)

mean < 1: There is data in the `product_category_name` column in the `products` dataset that is not in the `product_category_name` column in the `category_name` dataset. Here is a list of that data:

In [111]:
products.loc[
    ~products['product_category_name'].isin(category_name['product_category_name']),
    'product_category_name'
].unique()

array([nan, 'pc_gamer', 'portateis_cozinha_e_preparadores_de_alimentos'],
      dtype=object)

FK: `product_category_name`

## 7. Olist Sellers Dataset

In [112]:
sellers['seller_zip_code_prefix'].isin(geolocation['geolocation_zip_code_prefix']).mean()

np.float64(0.9977382875605816)

mean < 1: There is data in the `seller_zip_code_prefix` column in the `sellers` dataset that is not in the `geolocation_zip_code_prefix` column in the `geolocation` dataset. Here is a list of that data:

In [113]:
sellers.loc[
    ~sellers['seller_zip_code_prefix'].isin(geolocation['geolocation_zip_code_prefix']),
    'seller_zip_code_prefix'
].unique()

array([82040, 91901, 72580,  2285,  7412, 71551, 37708])

FK: `seller_zip_code_prefix`

**Table 1.** Summaries
| Table | Grain | Primary Key (PK) | Foreign Key (FK) | Other Columns |
|:-------:|:-------:|:------------------:|:------------------:|:---------------:|
|orders|1 order|order_id|customer_id|order_status, order_purchase_timestamp, order_approved_at, order_delivered_carrier_date, order_delivered_customer_date, order_estimated_delivery_date|
|order_items|1 item within an order|(order_id, order_item_id)|order_id, product_id, seller_id|shipping_limit_date, price, freight_value|
|order_payments|1 payment within an order|(order_id, payment_sequential)|order_id|payment_type, payment_installments, payment_value|
|order_reviews|a relationship between a review & an order|(order_id, review_id)|order_id|review_score, review_comment_title, review_comment_message, review_creation_date, review_answer_timestamp|
|customers|1 customer_id|customer_id|customer_zip_code_prefix|customer_unique_id, customer_city, customer_state|
|products|1 product|product_id|product_category_name|product_name_lenght, product_description_lenght, product_photos_qty, product_weight_g, product_length_cm, product_height_cm, product_width_cm|
|sellers|1 seller|seller_id|seller_zip_code_prefix|seller_city, seller_state|
|geolocation|-|-|-|-|
|category_name|1 product name|product_category_name|-|product_category_name_english|

# Join Logic Testing

## 1. Orders LEFT JOIN Customers

In [114]:
orders_customers = orders.merge(
    customers,
    on='customer_id',
    how='left'
)

In [115]:
orders_customers.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
2           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   
3           2017-12-15 00:00:00  7c142cf63193a1473d2e66489a9ae977   
4           2018-02-26 00:00:00  72632f0f9dd73dfee390c9b22eb56dd6   

   customer_zip_code_prefix            customer_city customer_state  
0                      3149                sao paulo             SP  
1                     47813                barreiras             BA  
2                     75265               vianopolis             GO  
3                     59296  sao goncalo do amarante             RN  
4                      9195              santo andre             SP

In [117]:
orders_customers.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
dtype: int64

All orders are placed by customers, and all customers place orders.

In [118]:
orders_customers.groupby('order_id')['customer_id'].nunique().count() == orders.shape[0]

np.True_

1 order only has 1 `customer_id`.

## 2. Orders LEFT JOIN Order Items

In [119]:
orders_items = orders.merge(
    order_items,
    on='order_id',
    how='left'
)

In [120]:
orders_items.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  order_item_id  \
0           2017-10-18 00:00:00            1.0   
1           2018-08-13 00:00:00            1.0   
2           2018-09-04 00:00:00            1.0   
3           2017-12-15 00:00:00            1.0   
4           2018-02-26 00:00:00            1.0   

                         product_id                         seller_id  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962   
2  aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2   
3  d0b61bfb1de832b15ba9d266ca96e5b0  66922902710d126a0e7d26b0e3805106   
4  65266b2da20d04dbe00c5c2d3bb7859e  2c9e548be18521d1c43cde1c582c6de8   

   shipping_limit_date   price  freight_value  
0  2017-10-06 11:07:15   29.99           8.72  
1  2018-07-30 03:24:27  118.70          22.76  
2  2018-08-13 08:55:23  159.90          19.22  
3  2017-11-23 19:45:59   45.00          27.20  
4  2018-02-19 20:31:37   19.90           8.72

In [121]:
orders_items.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 161
order_delivered_carrier_date     1968
order_delivered_customer_date    3229
order_estimated_delivery_date       0
order_item_id                     775
product_id                        775
seller_id                         775
shipping_limit_date               775
price                             775
freight_value                     775
dtype: int64

**Indicate:** There are orders without items (because `order_item_id` contains 775 null values).

## 3. Orders LEFT JOIN Order Payments

In [122]:
orders_payments = orders.merge(
    order_payments,
    on='order_id',
    how='left'
)

In [123]:
orders_payments.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  payment_sequential payment_type  \
0           2017-10-18 00:00:00                 1.0  credit_card   
1           2017-10-18 00:00:00                 3.0      voucher   
2           2017-10-18 00:00:00                 2.0      voucher   
3           2018-08-13 00:00:00                 1.0       boleto   
4           2018-09-04 00:00:00                 1.0  credit_card   

   payment_installments  payment_value  
0                   1.0          18.12  
1                   1.0           2.00  
2                   1.0          18.59  
3                   1.0         141.46  
4                   3.0         179.12

In [124]:
orders_payments.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 175
order_delivered_carrier_date     1888
order_delivered_customer_date    3132
order_estimated_delivery_date       0
payment_sequential                  1
payment_type                        1
payment_installments                1
payment_value                       1
dtype: int64

**Indicate:** There is order without payment (because `payment` columns contain one null value).

## 4. Orders LEFT JOIN Order Reviews

In [125]:
orders_reviews = orders.merge(
    order_reviews,
    on='order_id',
    how='left'
)

In [126]:
orders_reviews.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
2           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   
3           2017-12-15 00:00:00  359d03e676b3c069f62cadba8dd3f6e8   
4           2018-02-26 00:00:00  e50934924e227544ba8246aeb3770dd4   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0     Muito boa a loja   
2           5.0                  NaN   
3           5.0                  NaN   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1                               Muito bom o produto.  2018-08-08 00:00:00   
2                                                NaN  2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...  2017-12-03 00:00:00   
4                                                NaN  2018-02-17 00:00:00   

  review_answer_timestamp  
0     2017-10-12 03:43:48  
1     2018-08-08 18:37:50  
2     2018-08-22 19:07:58  
3     2017-12-05 19:21:58  
4     2018-02-18 13:02:51

In [127]:
orders_reviews.isna().sum()

order_id                             0
customer_id                          0
order_status                         0
order_purchase_timestamp             0
order_approved_at                  161
order_delivered_carrier_date      1793
order_delivered_customer_date     2987
order_estimated_delivery_date        0
review_id                          768
review_score                       768
review_comment_title             88424
review_comment_message           59015
review_creation_date               768
review_answer_timestamp            768
dtype: int64

**Indicate:** There are orders without reviews (because `review_id` column contain 768 null values).

## 5. Order Items LEFT JOIN Products

In [128]:
orders_items_products = order_items.merge(
    products,
    on='product_id',
    how='left'
)

In [129]:
orders_items_products.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value product_category_name  \
0  2017-09-19 09:45:35   58.90          13.29            cool_stuff   
1  2017-05-03 11:05:13  239.90          19.93              pet_shop   
2  2018-01-18 14:48:30  199.00          17.87      moveis_decoracao   
3  2018-08-15 10:10:18   12.99          12.79            perfumaria   
4  2017-02-13 13:57:51  199.90          18.14    ferramentas_jardim   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 58.0                       598.0                 4.0   
1                 56.0                       239.0                 2.0   
2                 59.0                       695.0                 2.0   
3                 42.0                       480.0                 1.0   
4                 59.0                       409.0                 1.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             650.0               28.0                9.0              14.0  
1           30000.0               50.0               30.0              40.0  
2            3050.0               33.0               13.0              33.0  
3             200.0               16.0               10.0              15.0  
4            3750.0               35.0               40.0              30.0

In [130]:
orders_items_products.isna().sum()

order_id                         0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
product_category_name         1603
product_name_lenght           1603
product_description_lenght    1603
product_photos_qty            1603
product_weight_g                18
product_length_cm               18
product_height_cm               18
product_width_cm                18
dtype: int64

In [131]:
orders_items_products['product_id'].duplicated().sum()

np.int64(79699)

In [132]:
orders_items_products[orders_items_products['order_item_id'].isin([2,3])]

order_id  order_item_id  \
14      0008288aa423d2a3f00fcb17cd7d8719              2   
33      00143d0f86d6fbd9f9b38ab440ac16f5              2   
34      00143d0f86d6fbd9f9b38ab440ac16f5              3   
43      001ab0a7578dd66cd4b0a71f5b6e1e41              2   
44      001ab0a7578dd66cd4b0a71f5b6e1e41              3   
...                                  ...            ...   
112616  ffecd5a79a0084f6a592288c67e3c298              2   
112617  ffecd5a79a0084f6a592288c67e3c298              3   
112635  fff8287bbae429a99bb7e8c21d151c41              2   
112641  fffb9224b6fc7c43ebb0904318b10b5f              2   
112642  fffb9224b6fc7c43ebb0904318b10b5f              3   

                              product_id                         seller_id  \
14      368c6c730842d78016ad823897a372db  1f50f920176fa81dab994f9023523100   
33      e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
34      e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
43      0b0172eb0fd18479d29c3bc122c058c2  5656537e588803a555b8eb41f07a944b   
44      0b0172eb0fd18479d29c3bc122c058c2  5656537e588803a555b8eb41f07a944b   
...                                  ...                               ...   
112616  50fd2b788dc166edd20512370dac54df  8b321bb669392f5163d04c59e235e066   
112617  50fd2b788dc166edd20512370dac54df  8b321bb669392f5163d04c59e235e066   
112635  bee2e070c39f3dd2f6883a17a5f0da45  4e922959ae960d389249c378d1c939f5   
112641  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112642  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   

        shipping_limit_date   price  freight_value   product_category_name  \
14      2018-02-21 02:55:52   49.90          13.37      ferramentas_jardim   
33      2017-10-20 16:07:52   21.33          15.10           esporte_lazer   
34      2017-10-20 16:07:52   21.33          15.10           esporte_lazer   
43      2018-01-04 02:33:42   24.89          17.63             eletronicos   
44      2018-01-04 02:33:42   24.89          17.63             eletronicos   
...                     ...     ...            ...                     ...   
112616  2018-03-05 20:15:27   21.90          15.79             eletronicos   
112617  2018-03-05 20:15:27   21.90          15.79             eletronicos   
112635  2018-03-27 12:29:22  180.00          48.14  informatica_acessorios   
112641  2017-11-03 02:55:58   55.00          34.19      relogios_presentes   
112642  2017-11-03 02:55:58   55.00          34.19      relogios_presentes   

        product_name_lenght  product_description_lenght  product_photos_qty  \
14                     57.0                       366.0                 2.0   
33                     52.0                      1153.0                 1.0   
34                     52.0                      1153.0                 1.0   
43                     53.0                       265.0                 2.0   
44                     53.0                       265.0                 2.0   
...                     ...                         ...                 ...   
112616                 46.0                      2285.0                 1.0   
112617                 46.0                      2285.0                 1.0   
112635                 53.0                       871.0                 4.0   
112641                 41.0                      1159.0                 4.0   
112642                 41.0                      1159.0                 4.0   

        product_weight_g  product_length_cm  product_height_cm  \
14                1650.0               30.0               22.0   
33                 180.0               17.0               11.0   
34                 180.0               17.0               11.0   
43                 150.0               22.0               13.0   
44                 150.0               22.0               13.0   
...                  ...                ...                ...   
112616             150.0              

**Indicate:** Different `order_item_id` can have same `product_id` in one `order_id`. There are no items without `product_id`.

## 6. Order Items LEFT JOIN Sellers

In [133]:
orders_items_sellers = order_items.merge(
    sellers,
    on='seller_id',
    how='left'
)

In [134]:
orders_items_sellers.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  seller_zip_code_prefix  \
0  2017-09-19 09:45:35   58.90          13.29                   27277   
1  2017-05-03 11:05:13  239.90          19.93                    3471   
2  2018-01-18 14:48:30  199.00          17.87                   37564   
3  2018-08-15 10:10:18   12.99          12.79                   14403   
4  2017-02-13 13:57:51  199.90          18.14                   87900   

     seller_city seller_state  
0  volta redonda           SP  
1      sao paulo           SP  
2  borda da mata           MG  
3         franca           SP  
4         loanda           PR

In [135]:
orders_items_sellers.isna().sum()

order_id                  0
order_item_id             0
product_id                0
seller_id                 0
shipping_limit_date       0
price                     0
freight_value             0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [136]:
orders_items_sellers['seller_id'].duplicated().sum()

np.int64(109555)

In [137]:
orders_items_sellers[orders_items_sellers['order_item_id'].isin([2,3])]

order_id  order_item_id  \
14      0008288aa423d2a3f00fcb17cd7d8719              2   
33      00143d0f86d6fbd9f9b38ab440ac16f5              2   
34      00143d0f86d6fbd9f9b38ab440ac16f5              3   
43      001ab0a7578dd66cd4b0a71f5b6e1e41              2   
44      001ab0a7578dd66cd4b0a71f5b6e1e41              3   
...                                  ...            ...   
112616  ffecd5a79a0084f6a592288c67e3c298              2   
112617  ffecd5a79a0084f6a592288c67e3c298              3   
112635  fff8287bbae429a99bb7e8c21d151c41              2   
112641  fffb9224b6fc7c43ebb0904318b10b5f              2   
112642  fffb9224b6fc7c43ebb0904318b10b5f              3   

                              product_id                         seller_id  \
14      368c6c730842d78016ad823897a372db  1f50f920176fa81dab994f9023523100   
33      e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
34      e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
43      0b0172eb0fd18479d29c3bc122c058c2  5656537e588803a555b8eb41f07a944b   
44      0b0172eb0fd18479d29c3bc122c058c2  5656537e588803a555b8eb41f07a944b   
...                                  ...                               ...   
112616  50fd2b788dc166edd20512370dac54df  8b321bb669392f5163d04c59e235e066   
112617  50fd2b788dc166edd20512370dac54df  8b321bb669392f5163d04c59e235e066   
112635  bee2e070c39f3dd2f6883a17a5f0da45  4e922959ae960d389249c378d1c939f5   
112641  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112642  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   

        shipping_limit_date   price  freight_value  seller_zip_code_prefix  \
14      2018-02-21 02:55:52   49.90          13.37                   15025   
33      2017-10-20 16:07:52   21.33          15.10                   18055   
34      2017-10-20 16:07:52   21.33          15.10                   18055   
43      2018-01-04 02:33:42   24.89          17.63                   72015   
44      2018-01-04 02:33:42   24.89          17.63                   72015   
...                     ...     ...            ...                     ...   
112616  2018-03-05 20:15:27   21.90          15.79                    1212   
112617  2018-03-05 20:15:27   21.90          15.79                    1212   
112635  2018-03-27 12:29:22  180.00          48.14                   12327   
112641  2017-11-03 02:55:58   55.00          34.19                   24440   
112642  2017-11-03 02:55:58   55.00          34.19                   24440   

                  seller_city seller_state  
14      sao jose do rio preto           SP  
33                   sorocaba           SP  
34                   sorocaba           SP  
43                   brasilia           DF  
44                   brasilia           DF  
...                       ...          ...  
112616              sao paulo           SP  
112617              sao paulo           SP  
112635                jacarei           SP  
112641            sao goncalo           RJ  
112642            sao goncalo           RJ  

[12090 rows x 10 columns]

**Indicate:** Different `order_item_id` can have same `seller_id` in one `order_id`. There are no items without `seller_id`.